File: src/lib/core/agent/claude.ts
import type { AgentRunner, AgentRunOptions, AgentRunResult } from "./types";

/** Turn a long assistant text block into a short progress line. */
function textStep(text: string): string {
  const clean = text.replace(/\s+/g, " ").trim();
  if (!clean) return "";
  return clean.length > 140 ? clean.slice(0, 137) + "…" : clean;
}

/** Turn a tool_use block into a friendly progress line. */
function toolStep(block: { name?: string; input?: Record<string, unknown> }): string {
  const name = block.name ?? "tool";
  if (name === "Bash") {
    const cmd = String(block.input?.command ?? "");
    if (/last30days/.test(cmd)) return "Searching Reddit, X, YouTube, HN, GitHub & the web…";
    if (/python|uv /.test(cmd)) return "Running research scripts…";
    return "Working…";
  }
  if (name === "WebSearch") return "Searching the web…";
  if (name === "Read" || name === "Glob" || name === "Grep") return "Reading results…";
  if (name === "Write") return "Saving findings…";
  return `Using ${name}…`;
}

/**
 * Env for the SDK with API keys removed so it always authenticates via the Claude
 * **subscription** (CLI login). The Agent SDK bills the paid API whenever
 * ANTHROPIC_API_KEY/ANTHROPIC_AUTH_TOKEN is present, so we drop them here.
 */
function subscriptionEnv(): Record<string, string> {
  const env = { ...(process.env as Record<string, string>) };
  delete env.ANTHROPIC_API_KEY;
  delete env.ANTHROPIC_AUTH_TOKEN;
  return env;
}

/**
 * Claude Agent SDK runner (subscription auth via Claude Code CLI). The SDK is
 * imported dynamically so it is an OPTIONAL dependency: if it isn't installed,
 * getAgentRunner() simply won't select this adapter. Skills under `.claude/skills`
 * are auto-discovered because settingSources includes 'project'.
 */
export class ClaudeAgentRunner implements AgentRunner {
  readonly name = "claude-agent";
  readonly supportsTools = true;

  static async isAvailable(): Promise<boolean> {
    try {
      await import("@anthropic-ai/claude-agent-sdk");
      return true;
    } catch {
      return false;
    }
  }

  async run(opts: AgentRunOptions): Promise<AgentRunResult> {
    const { prompt, system, skills = [], maxTurns = 40, onStep } = opts;
    let finalText = "";
    let ok = false;
    let error: string | undefined;

    try {
      const { query } = await import("@anthropic-ai/claude-agent-sdk");
      const q = query({
        prompt,
        options: {
          cwd: process.env.AGENT_CWD || process.cwd(),
          settingSources: ["project"],
          skills,
          allowedTools: ["Bash", "Read", "Write", "WebSearch", "Glob", "Grep"],
          permissionMode: "bypassPermissions",
          ...(process.env.CLAUDE_MODEL ? { model: process.env.CLAUDE_MODEL } : {}),
          ...(system ? { systemPrompt: system } : {}),
          env: subscriptionEnv(),
          maxTurns,
        },
      });

      for await (const msg of q) {
        if (msg.type === "assistant") {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          const content: any[] = (msg as any).message?.content ?? [];
          for (const block of content) {
            if (block?.type === "text" && block.text) {
              const s = textStep(block.text);
              if (s) onStep?.(s);
            } else if (block?.type === "tool_use") {
              onStep?.(toolStep(block));
            }
          }
        } else if (msg.type === "result") {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          const m = msg as any;
          ok = m.subtype === "success";
          finalText = typeof m.result === "string" ? m.result : "";
          if (!ok) error = m.subtype || "agent_error";
        }
      }
    } catch (e) {
      ok = false;
      error = (e as Error)?.message ?? String(e);
    }

    return { text: finalText, ok, error };
  }
}

File: worker_py/agent/claude.py
"""Claude Agent SDK runner (skills like ``last30days``). Port of ``agent/claude.ts``.

Skills under ``.claude/skills`` are auto-discovered because ``setting_sources``
includes ``"project"`` and ``cwd`` is the repo root.
"""

import os
import re
from typing import Any, Dict, List, Optional

from ..claude_sdk import (
    block_text,
    block_tool_name,
    build_options,
    is_available,
    iter_assistant_blocks,
    result_of,
    subscription_env,
)
from .base import AgentRunner, OnStep, RunResult


def text_step(text: str) -> str:
    """Turn a long assistant text block into a short progress line."""
    clean = re.sub(r"\s+", " ", text).strip()
    if not clean:
        return ""
    return clean[:137] + "…" if len(clean) > 140 else clean


def tool_step(name: str, inp: Dict[str, Any]) -> str:
    """Turn a tool_use block into a friendly progress line."""
    if name == "Bash":
        cmd = str((inp or {}).get("command", ""))
        if re.search(r"last30days", cmd):
            return "Searching Reddit, X, YouTube, HN, GitHub & the web…"
        if re.search(r"python|uv ", cmd):
            return "Running research scripts…"
        return "Working…"
    if name == "WebSearch":
        return "Searching the web…"
    if name in ("Read", "Glob", "Grep"):
        return "Reading results…"
    if name == "Write":
        return "Saving findings…"
    return f"Using {name}…"


class ClaudeAgentRunner(AgentRunner):
    name = "claude-agent"
    supports_tools = True

    @staticmethod
    async def is_available() -> bool:
        return await is_available()

    async def run(
        self,
        *,
        prompt: str,
        system: Optional[str] = None,
        skills: Optional[List[str]] = None,
        max_turns: int = 40,
        on_step: Optional[OnStep] = None,
    ) -> RunResult:
        final_text = ""
        ok = False
        error: Optional[str] = None

        try:
            from claude_agent_sdk import query  # type: ignore

            options = build_options(
                cwd=os.environ.get("AGENT_CWD") or os.getcwd(),
                setting_sources=["project"],
                skills=skills or [],
                allowed_tools=["Bash", "Read", "Write", "WebSearch", "Glob", "Grep"],
                permission_mode="bypassPermissions",
                model=os.environ.get("CLAUDE_MODEL"),
                system_prompt=system,
                env=subscription_env(),
                max_turns=max_turns,
            )

            async for msg in query(prompt=prompt, options=options):
                for block in iter_assistant_blocks(msg):
                    tool = block_tool_name(block)
                    if tool is not None:
                        if on_step:
                            on_step(tool_step(tool, getattr(block, "input", {}) or {}))
                        continue
                    text = block_text(block)
                    if text:
                        step = text_step(text)
                        if step and on_step:
                            on_step(step)
                res = result_of(msg)
                if res is not None:
                    ok, final_text, subtype = res
                    if not ok:
                        error = subtype or "agent_error"
        except Exception as e:  # noqa: BLE001
            ok = False
            error = str(e)

        return RunResult(text=final_text, ok=ok, error=error)
